lsp-mode.el

```diff
@@ -3702,77 +3702,101 @@ and the position respectively."
-         (when additional-text-edits
-           (lsp--apply-text-edits additional-text-edits)))
-
-       (when lsp-signature-auto-activate
-         (lsp-signature-activate))
-
-       (when (lsp--looking-back-trigger-characters-p trigger-chars)
-         (setq this-command 'self-insert-command))))))
+          ((eq action 'metadata)
+           `(metadata . ((display-sort-function . (lambda (candidates)
+                                                    (--sort (string-lessp (get-text-property 0 'lsp-sort-text it)
+                                                                          (get-text-property 0 'lsp-sort-text other))
+                                                            candidates))))))
+          ((eq (car-safe action) 'boundaries) nil)
+          (t (if done?
+                 result
+               (setf start-point (point))
+               (-let* ((resp (lsp-request-while-no-input "textDocument/completion"
+                                                         (plist-put (lsp--text-document-position-params)
+                                                                    :context (ht ("triggerKind" 1)))))
+                       (items (lsp--sort-completions (cond
+                                                      ((seqp resp) resp)
+                                                      ((hash-table-p resp) (gethash "items" resp)))))
+                       ((start . end) (or
+                                       (-when-let* ((item (lsp-elt items 0))
+                                                    ((&hash "textEdit" text-edit) item)
+                                                    ((&hash "range") text-edit))
+                                         (lsp--range-to-region range))
+                                       bounds))
+                       (probe
+                        (if (and start end) (buffer-substring-no-properties start end)
+                          "")))
+                 (setf done? (or (seqp resp)
+                                 (not (gethash "isIncomplete" resp)))
+                       result (--> (seq-into (--map (-let [(&hash "label"
+                                                                  "filterText" filter-text)
+                                                           it]
+                                                      (propertize (or filter-text label)
+                                                                  'lsp-completion-item it))
+                                                    items)
+                                             'list)
+                                   (completion-all-completions probe it pred (length probe))
+                                   (let ((tail (last it)))
+                                     (if (consp tail) (setcdr tail nil))
+                                     it)
+                                   (--map (-let* ((item (get-text-property 0 'lsp-completion-item it))
```

Comment:
Can you avoid using nested anamorphic forms(replace --map with -map)? Then make the code a bit harder to read since it changes it meanings in very narrow block of code.

Reply:
Done